### DISCUSSION TITLE
Trying getting TS004F  using all 4 EP with quirk

### DISCUSSION THREAD

**Discussion by MattWestb** [zigpy/zha-device-handlers#1316] [Ideas]
I'm trying getting the TS004F dimmer switch changing mode to "tuya scene switch".
The mode switch attribute is implanted and it can changing mode but it cant initiating the 3 new Endpoints. and is only using EP1.

For enabling them its needed having the device in INIT mode (first seconds after its joining) and writing the attribute to it before its leaving INIT mode and cant doing the "hardware" changes.

For getting the device in INIT mode tuya is sending one leave with rejoin flag.
That was not possible doing with ZHA but on fix is on the way and one work around is in place by first defining the flags:
```
class LeaveOptions(t.bitmap8):
    NONE = 0

    Rejoin = 0b10000000
    RemoveChildren = 0b01000000
```
And then in the device class replacement:
```
    #    replacement = signature  # this quirk does not change anything
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        # Immediately call `_leave_and_rejoin` when the quirk loads
        asyncio.create_task(self._leave_and_rejoin())

    async def _leave_and_rejoin(self):
        # Tell the device to leave and re-join
        await asyncio.sleep(10.0)
        await self.zdo.Mgmt_Leave_req(self.ieee, LeaveOptions.Rejoin)


        # Delete the device from the database

    #        self.application.listener_event("device_removed", self)

    # Delete the device from zigpy
    #        self.application.devices.pop(self.ieee, None)
```
This is working so then the device is joining and the quirk is loaded its wait 10 seconds and sending on leave with rejoin flag. Its looping sometimes until the device stop leaving.

Was trying writing attributes from Puddlys instruction:
```
class SomeEnum(t.enum8):
    SCENE_MODE = 0x01


ep = device.endpoints[???]

# 0xFFFE I think is `cluster_revision` in ZCL R7
await ep.basic.read_attributes(["manufacturer", "zcl_version", "app_version", "model", "power_source", 0xFFFE])
await asyncio.sleep(0.20)

# If the name of a cluster is not known, you can always look it up by its ID
await ep.in_clusters[0xE001].read_attributes([0xD011])
await asyncio.sleep(0.05)

await ep.power.read_attributes(["battery_voltage", "battery_percentage_remaining"])
await asyncio.sleep(0.05)

# If an attribute is defined on the cluster, you can also write to it by its name string
# instead of the attribute ID
await ep.on_off.write_attributes({0x8004: SomeEnum.SCENE_MODE})
await asyncio.sleep(0.05)

await device.endpoints[0x01].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x02].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x03].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x04].on_off.bind()
```
But i getting problem with device and EP. Was trying setting the EP to 1 but then the system is not getting the device.
Allos good is listening for commands like Puddlys example:
```
class NewCustomDevice(CustomDevice):
    signature = ...
    replacement = ...

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.add_context_listener(self.zdo)

    def zdo_device_annce(self, device, dst_addressing, hdr, args):
```
But the same problem with it i is missing to many imports and cant getting it working.

If some with more code knowledge can helping defining one cluster command on OnOff cluster (perhaps 0x44) that can being catched of the quirk and running one script that is doing:
- 1 Sending one leave with rejoining
- 2 Starting listening for` zdo_device_annce
- 3 Then getting one DC write one attribute to OnOff cluster
- 4 Reading some attribute
- 5 Reading how many endpoints is having.

Then i have enough pieces for getting the right timing for getting the device initiating the 3 new endpoints with little testing.

Attaching the last working in progress quirk.
[ts004f4.zip](https://github.com/zigpy/zha-device-handlers/files/7966128/ts004f4.zip)

Testing that the quirk is working OK on other device shall not being difficult its only need basic and OnOff cluster and all import and dependency is working OK so i can getting  the timing OK.

**Comment by MattWestb** [comment]
@jacekk015 You dont have the zuya switch but can you helping impanting one command 0x44 on the (tuya) OnOff cluster that is triggering one code block / class in the quirk so i can putting command for reading and writing commands and attribute from it ?
If getting one class that can being triggered its much easier doing test then deleting the device, reatrig HA and  pairing the device all the time.

And i think you can very easy implanting the other functions or helping my getting them in place with the right formatting and import in the quirk.

Thanks in advance !!!

Mattias W

**Reply by jacekk015** [reply]
I'm not sure what you want to achieve here. Does the device doesn't work properly??

**Reply by MattWestb** [reply]
The TS004F quirk in ZHA is working as default mode (as on dimmer switch with very strange cluster settings) and the attribute for switching the device to "scene mode" is implanted and working but its using only one EP (1) and the rest is not populated (cant doing the then the device is being initiated). 

As our Hubitat friens was writing https://github.com/zigpy/zha-device-handlers/issues/1286#issuecomment-1025399163 is it possible making the switch inisiating the 3 "new" EPs by reading and writing attributes to the device then its in INIT mode for very short time after it have being joining the network.

We have looking in tuya ZBGW paring and they is first paring the remote and then sending on leav with rejoin flag.
Then the its rejoining the system is reading some attributes and then wring one, and the remote is starting using the new endpoints. Then its possible deleting the device in ZHA and Zigbee. db and sending one more rejoining and ZHA is reading all new EPs and loading one new device class as one 4  EP device.

I need help getting function sending and reading attribute to EP1.
Also listening  for networks command sent by the device so can start the preces in the right  time for having the device in init mode.

The leave command i can sending with delay from the quirk is loaded but its no good way to knowing if the device is in init mode and need listening for device announcement and / or device timeout out request.

But for getting it working without getting the device configured every time its re/joining the network is shall being better implanting one cluster command or writing one attribute that the quirk is using for starting sending the leave command and then listening for rejoin and device time out request and doing the rest with normal attribute read and write.

So is it possible making one attribute write or cluster commands that the quirk is "catching" and start execute one or more conde classes ?

My idea is: ZHA mange cluster, Cluster OnOff (or power config) and command X or attribute X write 1 and the quirks is running one code like is done then one tuya TRVs is sending one DP command but its being sent from ZHA to the quirk.

Do you have some other good way triggering the quirk starting one code block and for sending commands ?

PS: Its looks that  the device is using 4 EP all the time then have getting in initiating it also after resetting the device but its losing its possibility if taking out the battery (like many other tuya device that have being paired with tuya ZBGW).

**Reply by jacekk015** [reply]
Are you saying that only one button works now??
Do you have a link for Aliexpress or somewhere for that device?

**Reply by MattWestb** [reply]
Default mode as dimmer its have 4 buttons (on. off dimm up and down).and one endpoint for sending and receiving commands.
Then switches to "scene" mode and getting initiated OK its using 4 endpoints one for etch button with short, double and long press.
If not  getting it in init mode the switching mode to scene mode is only EP initiated and can can only sending scene commands from first button = EP1.

The attribute for changing the switch mode:
https://github.com/zigpy/zha-device-handlers/blob/23a166902e50d5c32019e0d4d32a968715041bb6/zhaquirks/tuya/__init__.py#L836-L840

https://www.aliexpress.com/item/1005002181182397.html?spm=a2g0o.9042311.0.0.27424c4dEieMkr

**Reply by MattWestb** [reply]
Puddly was making one example of the command that shall being sent to the device after rejoining the network:
```
class SomeEnum(t.enum8):
    SCENE_MODE = 0x01


ep = device.endpoints[???]  # (I putting in 1 then its the right EP)

# 0xFFFE I think is `cluster_revision` in ZCL R7
await ep.basic.read_attributes(["manufacturer", "zcl_version", "app_version", "model", "power_source", 0xFFFE])
await asyncio.sleep(0.20)

# If the name of a cluster is not known, you can always look it up by its ID
await ep.in_clusters[0xE001].read_attributes([0xD011])
await asyncio.sleep(0.05)

await ep.power.read_attributes(["battery_voltage", "battery_percentage_remaining"])
await asyncio.sleep(0.05)

# If an attribute is defined on the cluster, you can also write to it by its name string
# instead of the attribute ID
await ep.on_off.write_attributes({0x8004: SomeEnum.SCENE_MODE})
await asyncio.sleep(0.05)

await device.endpoints[0x01].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x02].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x03].on_off.bind()
await asyncio.sleep(0.05)
await device.endpoints[0x04].on_off.bind()
```
I like putting it in the replacement part (after the cluster definitions) of the device quirk but i cant getting it linking to all system functions and is getting errors.

How to making one working class that is possible sending and reading cluster attributes from the device ?

Can you giving one example that making one class that is reading one attribute from one cluster on one EP ?
Then i can modifying it and extending it for getting the function working.
